safehttp/plugins/csp/csp.go

```diff
@@ -0,0 +1,201 @@
+	//
+	// See https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/base-uri
+	// for more info.
+	BaseURI string
+	// ReportURI controls the report-uri directive. If ReportUri is empty, no report-uri
+	// directive will be set.
+	ReportURI string
+}
+
+// Build creates a Policy based on the specified options.
+func (s StrictCSPBuilder) Build() Policy {
+	return Policy{
+		serialize: func(nonce string) string {
+			var b strings.Builder
+
+			// object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-{random}'
+			b.WriteString("object-src 'none'; script-src 'unsafe-inline' https: http: 'nonce-")
+			b.WriteString(nonce)
+			b.WriteByte('\'')
+
+			if s.StrictDynamic {
+				b.WriteString(" 'strict-dynamic'")
+			}
+			if s.UnsafeEval {
+				b.WriteString(" 'unsafe-eval'")
+			}
+
+			b.WriteString("; base-uri ")
+			if s.BaseURI == "" {
+				b.WriteString("'none'")
+			} else {
+				b.WriteString(s.BaseURI)
+			}
+
+			if s.ReportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(s.ReportURI)
+			}
+
+			return b.String()
+		},
+	}
+}
+
+// FramingPolicyBuilder can be used to create a new CSP policy with frame-ancestors
+// set to 'self'.
+//
+// TODO: allow relaxation on specific endpoints according to #77.
+type FramingPolicyBuilder struct {
+	// ReportURI controls the report-uri directive. If ReportUri is empty, no report-uri
+	// directive will be set.
+	ReportURI string
+}
+
+// Build creates a Policy based on the specified options.
+func (f FramingPolicyBuilder) Build() Policy {
+	return Policy{
+		serialize: func(_ string) string {
+			var b strings.Builder
+			b.WriteString("frame-ancestors 'self'")
+
+			if f.ReportURI != "" {
+				b.WriteString("; report-uri ")
+				b.WriteString(f.ReportURI)
+			}
+
+			return b.String()
+		},
+	}
+}
+
+// Interceptor intercepts requests and applies CSP policies.
+type Interceptor struct {
+	// Enforce specifies which policies will be set as the Content-Security-Policy
+	// header.
+	Enforce []Policy
+	// ReportOnly specifies which policies will be set as the Content-Security-Policy-Report-Only
+	// header.
+	ReportOnly []Policy
+}
+
+// Default creates a new CSP interceptor with a strict nonce-based policy and a
+// framing policy, both in enforcement mode.
+func Default(reportURI string) Interceptor {
+	return Interceptor{
+		Enforce: []Policy{
+			StrictCSPBuilder{ReportURI: reportURI}.Build(),
+			FramingPolicyBuilder{ReportURI: reportURI}.Build(),
+		},
+	}
+}
+
+// Before claims and sets the Content-Security-Policy header and the
+// Content-Security-Policy-Report-Only header.
+func (it Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	nonce := generateNonce()
+	r.SetContext(context.WithValue(r.Context(), ctxKey{}, nonce))
+
+	var csps []string
+	for _, p := range it.Enforce {
+		csps = append(csps, p.serialize(nonce))
+	}
+	var reportCsps []string
```

Comment:
You have `reportCsps` and `setCSP`, you could choose only one way of abbreviating it so the code is more uniform.